Segfault in complex finite difference stencil in multithreaded

I made a stencil generator for a 3D finite difference model that generates a complex stencil on a staggered grid. The usage of `@tturbo` results in a segfault in multithreaded mode, although the single thread performance is excellent. The code can be found on GitHub [here](https://github.com/microhh/stencilbuilder/blob/master/julia/stencil_julia_4th.jl)

Since this is my first Julia experience, I would like to learn whether this is a bug, or whether I ask too much from `LoopVectorization`.

Can I do something to contribute finding the cause? I have tested to split the stencil into smaller parts in this form:
```
    @tturbo for k in ks:ke
        for j in js:je
            for i in is:ie
                @fd (ut, u, v, w) ut += (
                    - gradx(interpx(u) * interpx(u)) + visc * (gradx(gradx(u))))
                @fd (ut, u, v, w) ut += (
                    - grady(interpx(v) * interpy(u)) + visc * (grady(grady(u))))
                @fd (ut, u, v, w) ut += (
                    - gradz(interpx(w) * interpz(u)) + visc * (gradz(gradz(u))) )
            end 
        end 
    end 
```
 but then I get 
```
ERROR: LoadError: LoadError: AssertionError: length(ops) ≤ 256
Stacktrace:
 [1] adjust_offsets!(ls::LoopVectorization.LoopSet, i::Int64, array_refs_with_same_name::Vector{Int64}, arrayref_to_name_op_collection::Vector{Vector{Tuple{Int64, Int64, Int64}}})
 ```

--- a/Project.toml
+++ b/Project.toml
@@ -1,7 +1,7 @@
 name = "LoopVectorization"
 uuid = "bdcacae8-1622-11e9-2a5c-532679323890"
 authors = ["Chris Elrod <elrodc@gmail.com>"]
-version = "0.12.54"
+version = "0.12.55"
 
 [deps]
 ArrayInterface = "4fba245c-0d91-5ea0-9b3e-6abc04ee57a9"
@@ -25,7 +25,7 @@ IfElse = "0.1"
 OffsetArrays = "1.4.1"
 Polyester = "0.3"
 Requires = "1"
-SLEEFPirates = "0.6.18"
+SLEEFPirates = "0.6.23"
 Static = "0.2, 0.3"
 StrideArraysCore = "0.1.12"
 ThreadingUtilities = "0.4.5"

--- a/src/broadcast.jl
+++ b/src/broadcast.jl
@@ -393,44 +393,44 @@ end
 @generated function vmaterialize!(
     dest::AbstractArray{T,N}, bc::BC, ::Val{Mod}, ::Val{UNROLL}
 ) where {T <: NativeTypes, N, BC <: Union{Broadcasted,Product}, Mod, UNROLL}
-    # 2+1
-    # we have an N dimensional loop.
-    # need to construct the LoopSet
-    # @show typeof(dest)
-    ls = LoopSet(Mod)
-    inline, u₁, u₂, isbroadcast, W, rs, rc, cls, l1, l2, l3, threads, warncheckarg = UNROLL
-    set_hw!(ls, rs, rc, cls, l1, l2, l3)
-    ls.isbroadcast = isbroadcast # maybe set `false` in a DiffEq-like `@..` macro
-    loopsyms = [gensym!(ls, "n") for n ∈ 1:N]
-    add_broadcast_loops!(ls, loopsyms, :dest)
-    elementbytes = sizeof(T)
-    add_broadcast!(ls, :dest, :bc, loopsyms, BC, elementbytes)
-    storeop = add_simple_store!(ls, :dest, ArrayReference(:dest, loopsyms), elementbytes)
-    doaddref!(ls, storeop)
-    resize!(ls.loop_order, num_loops(ls)) # num_loops may be greater than N, eg Product
+  # 2+1
+  # we have an N dimensional loop.
+  # need to construct the LoopSet
+  # @show typeof(dest)
+  ls = LoopSet(Mod)
+  inline, u₁, u₂, v, isbroadcast, W, rs, rc, cls, l1, l2, l3, threads, warncheckarg = UNROLL
+  set_hw!(ls, rs, rc, cls, l1, l2, l3)
+  ls.isbroadcast = isbroadcast # maybe set `false` in a DiffEq-like `@..` macro
+  loopsyms = [gensym!(ls, "n") for n ∈ 1:N]
+  add_broadcast_loops!(ls, loopsyms, :dest)
+  elementbytes = sizeof(T)
+  add_broadcast!(ls, :dest, :bc, loopsyms, BC, elementbytes)
+  storeop = add_simple_store!(ls, :dest, ArrayReference(:dest, loopsyms), elementbytes)
+  doaddref!(ls, storeop)
+  resize!(ls.loop_order, num_loops(ls)) # num_loops may be greater than N, eg Product
   # return ls
-    sc = setup_call(ls, :(Base.Broadcast.materialize!(dest, bc)), LineNumberNode(0), inline, false, u₁, u₂, threads%Int, warncheckarg)
+  sc = setup_call(ls, :(Base.Broadcast.materialize!(dest, bc)), LineNumberNode(0), inline, false, u₁, u₂, v, threads%Int, warncheckarg)
   # return sc
-    Expr(:block, Expr(:meta,:inline), sc, :dest)
+  Expr(:block, Expr(:meta,:inline), sc, :dest)
 end
 @generated function vmaterialize!(
     dest′::Union{Adjoint{T,A},Transpose{T,A}}, bc::BC, ::Val{Mod}, ::Val{UNROLL}
 ) where {T <: NativeTypes, N, A <: AbstractArray{T,N}, BC <: Union{Broadcasted,Product}, Mod, UNROLL}
-    # we have an N dimensional loop.
-    # need to construct the LoopSet
-    ls = LoopSet(Mod)
-    inline, u₁, u₂, isbroadcast, W, rs, rc, cls, l1, l2, l3, threads, warncheckarg = UNROLL
-    set_hw!(ls, rs, rc, cls, l1, l2, l3)
-    ls.isbroadcast = isbroadcast # maybe set `false` in a DiffEq-like `@..` macro
-    loopsyms = [gensym!(ls, "n") for n ∈ 1:N]
-    pushprepreamble!(ls, Expr(:(=), :dest, Expr(:call, :parent, :dest′)))
-    add_broadcast_loops!(ls, loopsyms, :dest′)
-    elementbytes = sizeof(T)
-    add_broadcast!(ls, :dest, :bc, loopsyms, BC, elementbytes)
-    storeop = add_simple_store!(ls, :dest, ArrayReference(:dest, reverse(loopsyms)), elementbytes)
-    doaddref!(ls, storeop)
-    resize!(ls.loop_order, num_loops(ls)) # num_loops may be greater than N, eg Product
-    Expr(:block, Expr(:meta,:inline), setup_call(ls, :(Base.Broadcast.materialize!(dest′, bc)), LineNumberNode(0), inline, false, u₁, u₂, threads%Int, warncheckarg), :dest′)
+  # we have an N dimensional loop.
+  # need to construct the LoopSet
+  ls = LoopSet(Mod)
+  inline, u₁, u₂, v, isbroadcast, W, rs, rc, cls, l1, l2, l3, threads, warncheckarg = UNROLL
+  set_hw!(ls, rs, rc, cls, l1, l2, l3)
+  ls.isbroadcast = isbroadcast # maybe set `false` in a DiffEq-like `@..` macro
+  loopsyms = [gensym!(ls, "n") for n ∈ 1:N]
+  pushprepreamble!(ls, Expr(:(=), :dest, Expr(:call, :parent, :dest′)))
+  add_broadcast_loops!(ls, loopsyms, :dest′)
+  elementbytes = sizeof(T)
+  add_broadcast!(ls, :dest, :bc, loopsyms, BC, elementbytes)
+  storeop = add_simple_store!(ls, :dest, ArrayReference(:dest, reverse(loopsyms)), elementbytes)
+  doaddref!(ls, storeop)
+  resize!(ls.loop_order, num_loops(ls)) # num_loops may be greater than N, eg Product
+  Expr(:block, Expr(:meta,:inline), setup_call(ls, :(Base.Broadcast.materialize!(dest′, bc)), LineNumberNode(0), inline, false, u₁, u₂, v, threads%Int, warncheckarg), :dest′)
 end
 # these are marked `@inline` so the `@turbo` itself can choose whether or not to inline.
 @generated function vmaterialize!(

--- a/src/codegen/lower_threads.jl
+++ b/src/codegen/lower_threads.jl
@@ -337,9 +337,9 @@ function scale_cost(c, looplen)
   c
 end
 function thread_one_loops_expr(
-    ls::LoopSet, ua::UnrollArgs, valid_thread_loop::Vector{Bool}, ntmax::UInt, c::Float64,
-    UNROLL::Tuple{Bool,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt}, OPS::Expr, ARF::Expr, AM::Expr, LPSYM::Expr
-  )
+  ls::LoopSet, ua::UnrollArgs, valid_thread_loop::Vector{Bool}, ntmax::UInt, c::Float64,
+  UNROLL::Tuple{Bool,Int8,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt}, OPS::Expr, ARF::Expr, AM::Expr, LPSYM::Expr
+)
   looplen = looplengthprod(ls)
   c = scale_cost(c, looplen)
   if all(isstaticloop, ls.loops)
@@ -473,9 +473,9 @@ function define_thread_blocks(threadedloop1, threadedloop2, vloop, u‚ÇÅloop, u‚Ç
   end
 end
 function thread_two_loops_expr(
-    ls::LoopSet, ua::UnrollArgs, valid_thread_loop::Vector{Bool}, ntmax::UInt, c::Float64,
-    UNROLL::Tuple{Bool,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt}, OPS::Expr, ARF::Expr, AM::Expr, LPSYM::Expr
-  )
+  ls::LoopSet, ua::UnrollArgs, valid_thread_loop::Vector{Bool}, ntmax::UInt, c::Float64,
+  UNROLL::Tuple{Bool,Int8,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt}, OPS::Expr, ARF::Expr, AM::Expr, LPSYM::Expr
+)
   looplen = looplengthprod(ls)
   # c = 0.0225 * c / looplen
   c = scale_cost(c, looplen)
@@ -677,7 +677,7 @@ function valid_thread_loops(ls::LoopSet)
   valid_thread_loop, ua, c
 end
 function avx_threads_expr(
-    ls::LoopSet, UNROLL::Tuple{Bool,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt},
+    ls::LoopSet, UNROLL::Tuple{Bool,Int8,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt},
     nt::UInt, OPS::Expr, ARF::Expr, AM::Expr, LPSYM::Expr
 )
     valid_thread_loop, ua, c = valid_thread_loops(ls)

--- a/src/codegen/lowering.jl
+++ b/src/codegen/lowering.jl
@@ -661,7 +661,7 @@ function push_outer_reduct_types!(pt::Expr, ls::LoopSet, ortypdefined::Bool)
   for j ∈ ls.outer_reductions
     oreducop = ls.operations[j]
     if ortypdefined
-      push!(pt.args, typeof_expr(oreducop))
+      push!(pt.args, eltype_expr(oreducop))
     else
       push!(pt.args, outer_reduct_init_typename(oreducop))
     end
@@ -679,7 +679,7 @@ function determine_eltype(ls::LoopSet, ortypdefined::Bool)::Union{Symbol,Expr}
     else
       oreducop = ls.operations[ls.outer_reductions[1]]
       if ortypdefined
-        return typeof_expr(oreducop)
+        return eltype_expr(oreducop)
       else
         return outer_reduct_init_typename(oreducop)
       end
@@ -886,18 +886,20 @@ function lower(ls::LoopSet, inline::Int = -1)
   order, u₁loop, u₂loop, vectorized, u₁, u₂, c, shouldinline = choose_order_cost(ls)
   lower(ls, order, u₁loop, u₂loop, vectorized, u₁, u₂, inlinedecision(inline, shouldinline))
 end
-function lower(ls::LoopSet, u₁::Int, u₂::Int, inline::Int)
+function lower(ls::LoopSet, u₁::Int, u₂::Int, v::Int, inline::Int)
   fill_offset_memop_collection!(ls)
   fill_children!(ls)
   if u₂ > 1
     @assert num_loops(ls) > 1 "There is only $(num_loops(ls)) loop, but specified blocking parameter u₂ is $u₂."
-    order, u₁loop, u₂loop, vectorized, _u₁, _u₂, c, shouldinline = choose_tile(ls)
+    order, u₁loop, u₂loop, vectorized, _u₁, _u₂, c, shouldinline = choose_tile(ls, store_load_deps(operations(ls)), v)
     copyto!(ls.loop_order.bestorder, order)
-  else
+  elseif u₁ > 0
     u₂ = -1
-    order, vectorized, c = choose_unroll_order(ls, Inf)
+    order, vectorized, c = choose_unroll_order(ls, Inf, store_load_deps(operations(ls)), v)
     u₁loop = first(order); u₂loop = Symbol("##undefined##"); shouldinline = true
     copyto!(ls.loop_order.bestorder, order)
+  else
+    order, u₁loop, u₂loop, vectorized, u₁, u₂, c, shouldinline = choose_order_cost(ls, v)
   end
   doinline = inlinedecision(inline, shouldinline)
   lower(ls, order, u₁loop, u₂loop, vectorized, u₁, u₂, doinline)

--- a/src/condense_loopset.jl
+++ b/src/condense_loopset.jl
@@ -138,10 +138,11 @@ struct OperationStruct <: AbstractLoopOperation
   loopdeps::UInt128
   reduceddeps::UInt128
   childdeps::UInt128
-  parents::UInt128
+  parents₀::UInt128
+  parents₁::UInt128
   node_type::OperationType
+  symid::UInt16
   array::UInt8
-  symid::UInt8
 end
 optype(os) = os.node_type
 
@@ -166,13 +167,21 @@ end
 loopdeps_uint(ls::LoopSet, op::Operation) = shifted_loopset(ls, loopdependencies(op))
 reduceddeps_uint(ls::LoopSet, op::Operation) = shifted_loopset(ls, reduceddependencies(op))
 childdeps_uint(ls::LoopSet, op::Operation) = shifted_loopset(ls, reducedchildren(op))
-function parents_uint(ls::LoopSet, op::Operation)
+function parents_uint(oppv::AbstractVector{Operation})
   p = zero(UInt128)
-  for parent ∈ parents(op)
-    p <<= 8
+  for parent ∈ oppv
+    p <<= 16
     p |= identifier(parent)
   end
   p
+end
+function parents_uint(op::Operation)
+  opv = parents(op)
+  N = length(opv)
+  @assert N ≤ 16
+  p0 = parents_uint(view(opv, 1:min(8,N)))
+  p1 = N > 8 ? parents_uint(view(opv, 9:N)) : zero(p0)
+  p0, p1
 end
 function recursively_set_parents_true!(x::Vector{Bool}, op::Operation)
   x[identifier(op)] && return nothing # don't redescend
@@ -199,16 +208,16 @@ function getroots!(rooted::Vector{Bool}, ls::LoopSet)
   return rooted
 end
 function OperationStruct!(varnames::Vector{Symbol}, ids::Vector{Int}, ls::LoopSet, op::Operation)
-    instr = instruction(op)
-    ld = loopdeps_uint(ls, op)
-    rd = reduceddeps_uint(ls, op)
-    cd = childdeps_uint(ls, op)
-    p = parents_uint(ls, op)
-    array = accesses_memory(op) ? findmatchingarray(ls, op.ref) : 0x00
-    ids[identifier(op)] = id = findindoradd!(varnames, name(op))
-    OperationStruct(
-        ld, rd, cd, p, op.node_type, array, id
-    )
+  instr = instruction(op)
+  ld = loopdeps_uint(ls, op)
+  rd = reduceddeps_uint(ls, op)
+  cd = childdeps_uint(ls, op)
+  p0, p1 = parents_uint(op)
+  array = accesses_memory(op) ? findmatchingarray(ls, op.ref) : 0x00
+  ids[identifier(op)] = id = findindoradd!(varnames, name(op))
+  OperationStruct(
+    ld, rd, cd, p0, p1, op.node_type, id, array
+  )
 end
 ## turn a LoopSet into a type object which can be used to reconstruct the LoopSet.
 
@@ -527,10 +536,10 @@ end
     ::Val{CNFARG}, ::StaticInt{W}, ::StaticInt{RS}, ::StaticInt{AR}, ::StaticInt{NT},
     ::StaticInt{CLS}, ::StaticInt{L1}, ::StaticInt{L2}, ::StaticInt{L3}
 ) where {CNFARG,W,RS,AR,CLS,L1,L2,L3,NT}
-  inline,u₁,u₂,BROADCAST,thread = CNFARG
+  inline,u₁,u₂,v,BROADCAST,thread = CNFARG
   nt = min(thread % UInt, NT % UInt)
-  t = Expr(:tuple, inline, u₁, u₂, BROADCAST, W, RS, AR, CLS, L1,L2,L3, nt)
-  length(CNFARG) == 6 && push!(t.args, last(CNFARG))
+  t = Expr(:tuple, inline, u₁, u₂, v, BROADCAST, W, RS, AR, CLS, L1, L2, L3, nt)
+  length(CNFARG) == 7 && push!(t.args, CNFARG[7])
   Expr(:call, Expr(:curly, :Val, t))
 end
 @inline function avx_config_val(
@@ -563,7 +572,8 @@ end
 
 
 function split_ifelse!(
-  ls::LoopSet, preserve::Vector{Symbol}, shouldindbyind::Vector{Bool}, roots::Vector{Bool}, extra_args::Expr, k::Int, inlineu₁u₂::Tuple{Bool,Int8,Int8}, thread::UInt, debug::Bool
+  ls::LoopSet, preserve::Vector{Symbol}, shouldindbyind::Vector{Bool}, roots::Vector{Bool}, extra_args::Expr, k::Int,
+  inlineu₁u₂::Tuple{Bool,Int8,Int8,Int8}, thread::UInt, debug::Bool
 )
   roots[k] = false
   op = operations(ls)[k]
@@ -617,13 +627,14 @@ function split_ifelse!(
   prepre
 end
 
-function generate_call(ls::LoopSet, inlineu₁u₂::Tuple{Bool,Int8,Int8}, thread::UInt, debug::Bool)
+function generate_call(ls::LoopSet, inlineu₁u₂::Tuple{Bool,Int8,Int8,Int8}, thread::UInt, debug::Bool)
   extra_args = Expr(:tuple)
   preserve, shouldindbyind, roots = add_grouped_strided_pointer!(extra_args, ls)
   generate_call_split(ls, preserve, shouldindbyind, roots, extra_args, inlineu₁u₂, thread, debug)
 end
 function generate_call_split(
-  ls::LoopSet, preserve::Vector{Symbol}, shouldindbyind::Vector{Bool}, roots::Vector{Bool}, extra_args::Expr, inlineu₁u₂::Tuple{Bool,Int8,Int8}, thread::UInt, debug::Bool
+  ls::LoopSet, preserve::Vector{Symbol}, shouldindbyind::Vector{Bool}, roots::Vector{Bool}, extra_args::Expr,
+  inlineu₁u₂::Tuple{Bool,Int8,Int8,Int8}, thread::UInt, debug::Bool
 )
   for (k,op) ∈ enumerate(operations(ls))
     parents_op = parents(op)
@@ -636,7 +647,8 @@ end
 
 # Try to condense in type stable manner
 function generate_call_types(
-  ls::LoopSet, preserve::Vector{Symbol}, shouldindbyind::Vector{Bool}, roots::Vector{Bool}, extra_args::Expr, (inline,u₁,u₂)::Tuple{Bool,Int8,Int8}, thread::UInt, debug::Bool
+  ls::LoopSet, preserve::Vector{Symbol}, shouldindbyind::Vector{Bool}, roots::Vector{Bool}, extra_args::Expr,
+  (inline,u₁,u₂,v)::Tuple{Bool,Int8,Int8,Int8}, thread::UInt, debug::Bool
 )
   # good place to check for split  
   operation_descriptions = Expr(:tuple)
@@ -665,7 +677,7 @@ function generate_call_types(
   loop_syms = tuple_expr(QuoteNode, ls.loopsymbols)
   func = debug ? lv(:_turbo_loopset_debug) : lv(:_turbo_!)
   lbarg = debug ? Expr(:call, :typeof, loop_bounds) : loop_bounds
-  configarg = (inline,u₁,u₂,ls.isbroadcast,thread)
+  configarg = (inline,u₁,u₂,v,ls.isbroadcast,thread)
   unroll_param_tup = Expr(:call, lv(:avx_config_val), :(Val{$configarg}()), VECTORWIDTHSYMBOL)
   q = Expr(:call, func, unroll_param_tup, val(operation_descriptions), val(arrayref_descriptions), val(argmeta), val(loop_syms))
 
@@ -697,9 +709,10 @@ function generate_call_types(
 end
 # @inline reductinittype(::T) where {T} = StaticType{T}()
 typeof_expr(op::Operation) = Expr(:call, GlobalRef(Base,:typeof), name(op))
+eltype_expr(op::Operation) = Expr(:call, GlobalRef(Base,:eltype), name(op))
 function add_outerreduct_types!(extra_args::Expr, ls::LoopSet) # extract_outerreduct_types!
   for or ∈ ls.outer_reductions
-    push!(extra_args.args, typeof_expr(operations(ls)[or]))
+    push!(extra_args.args, eltype_expr(operations(ls)[or]))
   end
 end
 """
@@ -735,6 +748,7 @@ Returns true if the element type is supported.
 """
 @inline check_type(::Type{T}) where {T <: NativeTypes} = true
 @inline check_type(::Type{T}) where {T} = false
+@inline check_type(::Type{T}) where {T <: AbstractSIMD} = true
 @inline check_device(::ArrayInterface.CPUPointer) = true
 @inline check_device(::ArrayInterface.CPUTuple) = true
 @inline check_device(x) = false
@@ -787,10 +801,10 @@ function setup_call_final(ls::LoopSet, q::Expr)
   return ls.preamble
 end
 function setup_call_debug(ls::LoopSet)
-  generate_call(ls, (false,zero(Int8),zero(Int8)), zero(UInt), true)
+  generate_call(ls, (false,zero(Int8),zero(Int8),zero(Int8)), zero(UInt), true)
 end
 function setup_call(
-  ls::LoopSet, q::Expr, source::LineNumberNode, inline::Bool, check_empty::Bool, u₁::Int8, u₂::Int8, thread::Int, warncheckarg::Int
+  ls::LoopSet, q::Expr, source::LineNumberNode, inline::Bool, check_empty::Bool, u₁::Int8, u₂::Int8, v::Int8, thread::Int, warncheckarg::Int
 )
   # We outline/inline at the macro level by creating/not creating an anonymous function.
   # The old API instead was based on inlining or not inline the generated function, but
@@ -799,7 +813,7 @@ function setup_call(
   # inlining the generated function into the loop preamble.
   lnns = extract_all_lnns(q)
   pushfirst!(lnns, source)
-  call = generate_call(ls, (inline, u₁, u₂), thread%UInt, false)
+  call = generate_call(ls, (inline, u₁, u₂, v), thread%UInt, false)
   call = check_empty ? check_if_empty(ls, call) : call
   argfailure = make_crashy(make_fast(q))
   if warncheckarg ≠ 0

--- a/src/constructors.jl
+++ b/src/constructors.jl
@@ -35,28 +35,28 @@ function add_ci_call!(q::Expr, @nospecialize(f), args, syms, i, valarg = nothing
     push!(q.args, Expr(:(=), syms[i], call))
 end
 
-function substitute_broadcast(q::Expr, mod::Symbol, inline::Bool, u₁::Int8, u₂::Int8, threads::Int, warncheckarg::Int)
-    ci = first(Meta.lower(LoopVectorization, q).args).code
-    nargs = length(ci)-1
-    ex = Expr(:block)
-    syms = [gensym() for _ ∈ 1:nargs]
-    configarg = (inline,u₁,u₂,true,threads,warncheckarg)
-    unroll_param_tup = Expr(:call, lv(:avx_config_val), :(Val{$configarg}()), staticexpr(0))
-    for n ∈ 1:nargs
-        ciₙ = ci[n]
-        ciₙargs = ciₙ.args
-        f = first(ciₙargs)
-        if ciₙ.head === :(=)
-            push!(ex.args, Expr(:(=), f, syms[((ciₙargs[2])::Core.SSAValue).id]))
-        elseif isglobalref(f, Base, :materialize!)
-            add_ci_call!(ex, lv(:vmaterialize!), ciₙargs, syms, n, unroll_param_tup, mod)
-        elseif isglobalref(f, Base, :materialize)
-            add_ci_call!(ex, lv(:vmaterialize), ciₙargs, syms, n, unroll_param_tup, mod)
-        else
-            add_ci_call!(ex, f, ciₙargs, syms, n)
-        end
+function substitute_broadcast(q::Expr, mod::Symbol, inline::Bool, u₁::Int8, u₂::Int8, v::Int8, threads::Int, warncheckarg::Int)
+  ci = first(Meta.lower(LoopVectorization, q).args).code
+  nargs = length(ci)-1
+  ex = Expr(:block)
+  syms = [gensym() for _ ∈ 1:nargs]
+  configarg = (inline,u₁,u₂,v,true,threads,warncheckarg)
+  unroll_param_tup = Expr(:call, lv(:avx_config_val), :(Val{$configarg}()), staticexpr(0))
+  for n ∈ 1:nargs
+    ciₙ = ci[n]
+    ciₙargs = ciₙ.args
+    f = first(ciₙargs)
+    if ciₙ.head === :(=)
+      push!(ex.args, Expr(:(=), f, syms[((ciₙargs[2])::Core.SSAValue).id]))
+    elseif isglobalref(f, Base, :materialize!)
+      add_ci_call!(ex, lv(:vmaterialize!), ciₙargs, syms, n, unroll_param_tup, mod)
+    elseif isglobalref(f, Base, :materialize)
+      add_ci_call!(ex, lv(:vmaterialize), ciₙargs, syms, n, unroll_param_tup, mod)
+    else
+      add_ci_call!(ex, f, ciₙargs, syms, n)
     end
-    esc(ex)
+  end
+  esc(ex)
 end
 
 
@@ -75,7 +75,7 @@ function loopset(q::Expr) # for interactive use only
     ls
 end
 
-function check_macro_kwarg(arg, inline::Bool, check_empty::Bool, u₁::Int8, u₂::Int8, threads::Int, warncheckarg::Int)
+function check_macro_kwarg(arg, inline::Bool, check_empty::Bool, u₁::Int8, u₂::Int8, v::Int8, threads::Int, warncheckarg::Int)
     ((arg.head === :(=)) && (length(arg.args) == 2)) || throw(ArgumentError("macro kwarg should be of the form `argname = value`."))
     kw = (arg.args[1])::Symbol
     value = (arg.args[2])
@@ -90,6 +90,8 @@ function check_macro_kwarg(arg, inline::Bool, check_empty::Bool, u₁::Int8, u�
         else
             throw(ArgumentError("Don't know how to process argument in `unroll=$value`."))
         end
+    elseif kw === :vectorize
+        v = convert(Int8, value)
     elseif kw === :check_empty
         check_empty = value::Bool
     elseif kw === :thread
@@ -105,23 +107,23 @@ function check_macro_kwarg(arg, inline::Bool, check_empty::Bool, u₁::Int8, u�
     else
         throw(ArgumentError("Received unrecognized keyword argument $kw. Recognized arguments include:\n`inline`, `unroll`, `check_empty`, and `thread`."))
     end
-    inline, check_empty, u₁, u₂, threads, warncheckarg
+    inline, check_empty, u₁, u₂, v, threads, warncheckarg
 end
-function process_args(args; inline::Bool = false, check_empty::Bool = false, u₁::Int8 = zero(Int8), u₂::Int8 = zero(Int8), threads::Int = 1, warncheckarg::Int = 0)
+function process_args(args; inline::Bool = false, check_empty::Bool = false, u₁::Int8 = zero(Int8), u₂::Int8 = zero(Int8), v::Int8 = zero(Int8), threads::Int = 1, warncheckarg::Int = 0)
   for arg ∈ args
-    inline, check_empty, u₁, u₂, threads, warncheckarg = check_macro_kwarg(arg, inline, check_empty, u₁, u₂, threads, warncheckarg)
+    inline, check_empty, u₁, u₂, v, threads, warncheckarg = check_macro_kwarg(arg, inline, check_empty, u₁, u₂, v, threads, warncheckarg)
   end
-  inline, check_empty, u₁, u₂, threads, warncheckarg
+  inline, check_empty, u₁, u₂, v, threads, warncheckarg
 end
 function turbo_macro(mod, src, q, args...)
   q = macroexpand(mod, q)
   if q.head === :for
     ls = LoopSet(q, mod)
-    inline, check_empty, u₁, u₂, threads, warncheckarg = process_args(args)
-    esc(setup_call(ls, q, src, inline, check_empty, u₁, u₂, threads, warncheckarg))
+    inline, check_empty, u₁, u₂, v, threads, warncheckarg = process_args(args)
+    esc(setup_call(ls, q, src, inline, check_empty, u₁, u₂, v, threads, warncheckarg))
   else
-    inline, check_empty, u₁, u₂, threads, warncheckarg = process_args(args, inline=true)
-    substitute_broadcast(q, Symbol(mod), inline, u₁, u₂, threads, warncheckarg)
+    inline, check_empty, u₁, u₂, v, threads, warncheckarg = process_args(args, inline=true)
+    substitute_broadcast(q, Symbol(mod), inline, u₁, u₂, v, threads, warncheckarg)
   end
 end
 """
@@ -218,7 +220,7 @@ Setting it to an integer > 0 will warn that many times, while setting it to a ne
 an unlimited amount of times. The default is `warn_check_args = 0`.
 """
 macro turbo(args...)
-    turbo_macro(__module__, __source__, last(args), Base.front(args)...)
+  turbo_macro(__module__, __source__, last(args), Base.front(args)...)
 end
 """
     @tturbo
@@ -257,11 +259,11 @@ end
 macro _turbo(arg, q)
   @assert q.head === :for
   q = macroexpand(__module__, q)
-  inline, check_empty, u₁, u₂ = check_macro_kwarg(arg, false, false, zero(Int8), zero(Int8), 1, 0)
+  inline, check_empty, u₁, u₂, v = check_macro_kwarg(arg, false, false, zero(Int8), zero(Int8), zero(Int8), 1, 0)
   ls = LoopSet(q, __module__)
   set_hw!(ls)
   def_outer_reduct_types!(ls)
-  esc(Expr(:block, ls.prepreamble, lower(ls, u₁ % Int, u₂ % Int, -1)))
+  esc(Expr(:block, ls.prepreamble, lower(ls, u₁ % Int, u₂ % Int, v % Int, -1)))
 end
 
 """

--- a/src/modeling/determinestrategy.jl
+++ b/src/modeling/determinestrategy.jl
@@ -1254,14 +1254,16 @@ function reject_reorder(ls::LoopSet, loop::Symbol, isu₂::Bool)
   false
 end
 
-function choose_unroll_order(ls::LoopSet, lowest_cost::Float64 = Inf, sld::Vector{Vector{Symbol}} = store_load_deps(operations(ls)))
+function choose_unroll_order(ls::LoopSet, lowest_cost::Float64 = Inf, sld::Vector{Vector{Symbol}} = store_load_deps(operations(ls)), v::Int = 0)
   iszero(length(offsetloadcollection(ls).opidcollectionmap)) && fill_offset_memop_collection!(ls)
   lo = LoopOrders(ls)
   new_order, state = iterate(lo) # right now, new_order === best_order
+  vec_iter_prealloc = v == 0 ? new_order : [(ls.loops)[v].itersymbol]
   best_order = similar(new_order)
   best_vec = first(new_order)
   while true
-    for new_vec ∈ new_order
+    vec_iter = v == 0 ? new_order : vec_iter_prealloc
+    for new_vec ∈ vec_iter
       reject_reorder(ls, new_vec, false) && continue
       cost_temp = evaluate_cost_unroll(ls, new_order, new_vec, lowest_cost, sld)
       if cost_temp < lowest_cost
@@ -1336,10 +1338,11 @@ function reject_candidate(ls::LoopSet, u₁loopsym::Symbol, u₂loopsym::Symbol)
 end
 inlinedecision(inline::Int, shouldinline::Bool) = iszero(inline) ? shouldinline : isone(inline)
 
-function choose_tile(ls::LoopSet, sld::Vector{Vector{Symbol}} = store_load_deps(operations(ls)))
+function choose_tile(ls::LoopSet, sld::Vector{Vector{Symbol}} = store_load_deps(operations(ls)), v::Int = 0)
   iszero(length(offsetloadcollection(ls).opidcollectionmap)) && fill_offset_memop_collection!(ls)
   lo = LoopOrders(ls)
   best_order = ls.loop_order.bestorder
+  vec_iter_prealloc = v == 0 ? best_order : [(ls.loops)[v].itersymbol]
   bestu₁ = bestu₂ = best_vec = Symbol("")
   u₁ = u₂ = 0; lowest_cost = Inf; loadelim = false
   anyisbit = any(ls.loopindexesbit)
@@ -1354,7 +1357,8 @@ function choose_tile(ls::LoopSet, sld::Vector{Vector{Symbol}} = store_load_deps(
       (((newu₁ == newu₂) || reject_reorder(ls, newu₁, true)) || reject_candidate(ls, newu₁, newu₂)) && continue
       new_order, state = iterate(lo) # right now, new_order === best_order
       while true
-        for new_vec ∈ new_order # view to skip first
+        vec_iter = v == 0 ? new_order : vec_iter_prealloc
+        for new_vec ∈ vec_iter
           reject_reorder(ls, new_vec, false) && continue
           if anyisbit && ls.loopindexesbit[getloopid(ls,new_vec)]
             # ((new_vec === newu₁) || (new_vec === newu₂)) || continue
@@ -1405,19 +1409,19 @@ function mismatchedstorereductions(ls::LoopSet)
   false
 end
 # Last in order is the inner most loop
-function choose_order_cost(ls::LoopSet)
+function choose_order_cost(ls::LoopSet, v::Int = 0)
   fill_children!(ls)
   resize!(ls.loop_order, length(ls.loopsymbols))
   sld = store_load_deps(operations(ls))
   if num_loops(ls) > 1
-    torder, tunroll, ttile, tvec, tU, tT, tc, shouldinline = choose_tile(ls, sld)
+    torder, tunroll, ttile, tvec, tU, tT, tc, shouldinline = choose_tile(ls, sld, v)
   else
     torder = names(ls) # dummy
     tunroll = ttile = tvec = Symbol("##undefined##") # dummy
     tU = tT = 0 # dummy
     tc = Inf
   end
-  uorder, uvec, uc = choose_unroll_order(ls, tc, sld)
+  uorder, uvec, uc = choose_unroll_order(ls, tc, sld, v)
   mismatched = mismatchedstorereductions(ls)
   if num_loops(ls) > 1 && tc ≤ uc
     @assert ls.loop_order.bestorder === torder

--- a/src/reconstruct_loopset.jl
+++ b/src/reconstruct_loopset.jl
@@ -377,15 +377,15 @@ function expandbyoffset!(indexpand::Vector{T}, inds, offsets::Vector{Int}, expan
     indexpand
 end
 expandbyoffset(inds::Vector{Int}, offsets::Vector{Int}, expand::Bool) = expandbyoffset!(Int[], inds, offsets, expand)
-function loopindex(ls::LoopSet, u::Unsigned, shift::Unsigned)
-    mask = (one(shift) << shift) - one(shift) # mask to zero out all but shift-bits
-    idxs = Int[]
-    while u != zero(u)
-        pushfirst!(idxs, ( u % UInt8 ) & mask)
-        u >>= shift
-    end
-    reverse!(idxs)
+function loopindex!(idxs::Vector{Int}, ls::LoopSet, u::Unsigned, shift::Unsigned)
+  mask = (one(shift) << shift) - one(shift) # mask to zero out all but shift-bits
+  while u != zero(u)
+    pushfirst!(idxs, ( u % UInt8 ) & mask)
+    u >>= shift
+  end
+  idxs
 end
+loopindex(ls::LoopSet, u::Unsigned, shift::Unsigned) = reverse!(loopindex!(Int[], ls, u, shift))
 function loopindexoffset(ls::LoopSet, u::Unsigned, li::Bool, expand::Bool = false)
     if li
         shift = 0x04
@@ -406,8 +406,13 @@ reduceddependencies(ls::LoopSet, os::OperationStruct, expand = false, offset = 0
 childdependencies(ls::LoopSet, os::OperationStruct, expand = false, offset = 0) = parents_symvec(ls, os.childdeps, expand, offset)
 
 # parents(ls::LoopSet, u::UInt128) = loopindexoffset(ls, u, false)
-parents(ls::LoopSet, u::UInt128) = loopindex(ls, u, 0x08)
-parents(ls::LoopSet, os::OperationStruct) = parents(ls, os.parents)
+function parents(ls::LoopSet, u₀::UInt128, u₁::UInt128)
+  idxs = Int[]
+  u₁ == zero(u₁) || loopindex!(idxs, ls, u₁, 0x10)
+  loopindex!(idxs, ls, u₀, 0x10)
+  reverse!(idxs)
+end
+parents(ls::LoopSet, os::OperationStruct) = parents(ls, os.parents₀, os.parents₁)
 
 expandedopname(opsymbol::Symbol, offset::Integer) = Symbol(String(opsymbol)*'#'*string(offset+1)*'#')
 function calcnops(ls::LoopSet, os::OperationStruct)
@@ -477,13 +482,13 @@ function add_op!(
     push!(opoffsets, opoffsets[end] + nops)
     nothing
 end
-function add_parents_to_op!(ls::LoopSet, op::Operation, up::Unsigned, k::Int, Δ::Int)
+function add_parents_to_op!(ls::LoopSet, op::Operation, up₀::UInt128, up₁::UInt128, k::Int, Δ::Int)
     vparents = parents(op)
     ops = operations(ls)
     offsets = ls.operation_offsets
     if isone(Δ) # not expanded
         @assert isone(k)
-        for i ∈ parents(ls, up)
+        for i ∈ parents(ls, up₀, up₁)
             # FIXME; children also filled in cacheunrolled
             for j ∈ offsets[i]+1:offsets[i+1] # if parents are expanded, add them all
                 opp = ops[j]
@@ -495,7 +500,7 @@ function add_parents_to_op!(ls::LoopSet, op::Operation, up::Unsigned, k::Int, Δ
         # Do we want to require that all Δidxs are equal?
         # Because `CartesianIndex((2,3)) - 1` results in a methoderorr, I think this is reasonable for now
         # FIXME; children also filled in cacheunrolled
-        for i ∈ parents(ls, up)
+        for i ∈ parents(ls, up₀, up₁)
             opp = ops[offsets[i]+k]
             pushfirst!(vparents, opp)
             push!(children(opp), op)
@@ -516,7 +521,7 @@ function add_parents_to_ops!(ls::LoopSet, ops::Vector{OperationStruct}, constoff
                     pushpreamble!(ls, Expr(:(=), instr.instr, extract_varg(constoffset)))
                 end
             elseif !isloopvalue(op)
-                add_parents_to_op!(ls, op, ops[i].parents, k, Δ)
+                add_parents_to_op!(ls, op, ops[i].parents₀, ops[i].parents₁, k, Δ)
             end
         end
     end
@@ -637,9 +642,9 @@ function avx_loopset!(
     extractind = extract_outerreduct_types!(ls, extractind, vargs)
     ls
 end
-function avx_body(ls::LoopSet, UNROLL::Tuple{Bool,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt})
-    inline, u₁, u₂, isbroadcast, W, rs, rc, cls, l1, l2, l3, nt = UNROLL
-    q = iszero(u₁) ? lower_and_split_loops(ls, inline % Int) : lower(ls, u₁ % Int, u₂ % Int, inline % Int)
+function avx_body(ls::LoopSet, UNROLL::Tuple{Bool,Int8,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt})
+    inline, u₁, u₂, v, isbroadcast, W, rs, rc, cls, l1, l2, l3, nt = UNROLL
+    q = (iszero(u₁) & iszero(v)) ? lower_and_split_loops(ls, inline % Int) : lower(ls, u₁ % Int, u₂ % Int, v % Int, inline % Int)
     ls.isbroadcast = isbroadcast
     iszero(length(ls.outer_reductions)) ? push!(q.args, nothing) : push!(q.args, loopset_return_value(ls, Val(true)))
     q
@@ -662,14 +667,14 @@ function tovector(@nospecialize(t))
     v
 end
 function _turbo_loopset(
-    @nospecialize(OPSsv), @nospecialize(ARFsv), @nospecialize(AMsv), @nospecialize(LPSYMsv), LBsv::Core.SimpleVector, vargs::Core.SimpleVector,
-    UNROLL::Tuple{Bool,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt}
+  @nospecialize(OPSsv), @nospecialize(ARFsv), @nospecialize(AMsv), @nospecialize(LPSYMsv), LBsv::Core.SimpleVector, vargs::Core.SimpleVector,
+  UNROLL::Tuple{Bool,Int8,Int8,Int8,Bool,Int,Int,Int,Int,Int,Int,Int,UInt}
 )
     nops = length(OPSsv) ÷ 3
     instr = Instruction[Instruction(OPSsv[3i+1], OPSsv[3i+2]) for i ∈ 0:nops-1]
     ops = OperationStruct[ OPSsv[3i] for i ∈ 1:nops ]
     ls = LoopSet(:LoopVectorization)
-    inline, u₁, u₂, isbroadcast, W, rs, rc, cls, l1, l2, l3, nt = UNROLL
+    inline, u₁, u₂, v, isbroadcast, W, rs, rc, cls, l1, l2, l3, nt = UNROLL
     set_hw!(ls, rs, rc, cls, l1, l2, l3); ls.vector_width = W; ls.isbroadcast = isbroadcast
     arsv = Vector{ArrayRefStruct}(undef, length(ARFsv))
     for i ∈ eachindex(arsv)
@@ -711,10 +716,10 @@ Execute an `@turbo` block. The block's code is represented via the arguments:
   pushfirst!(ls.preamble.args, :(var"#lv#tuple#args#" = reassemble_tuple(Tuple{var"#LB#",var"#V#"}, var"#flattened#var#arguments#")))
   # return @show avx_body(ls, var"#UNROLL#")
   if last(var"#UNROLL#") > 1
-    inline, u₁, u₂, isbroadcast, W, rs, rc, cls, l1, l2, l3, nt = var"#UNROLL#"
+    inline, u₁, u₂, v, isbroadcast, W, rs, rc, cls, l1, l2, l3, nt = var"#UNROLL#"
     # wrap in `var"#OPS#", var"#ARF#", var"#AM#", var"#LPSYM#"` in `Expr` to homogenize types
     avx_threads_expr(
-      ls, (inline, u₁, u₂, isbroadcast, W, rs, rc, cls, l1, l2, l3, one(UInt)), nt,
+      ls, (inline, u₁, u₂, v, isbroadcast, W, rs, rc, cls, l1, l2, l3, one(UInt)), nt,
       :(Val{$(var"#OPS#")}()), :(Val{$(var"#ARF#")}()), :(Val{$(var"#AM#")}()), :(Val{$(var"#LPSYM#")}())
     )
   else

--- a/test/shuffleloadstores.jl
+++ b/test/shuffleloadstores.jl
@@ -15,6 +15,15 @@ function cdot_mat(ca::AbstractVector{Complex{T}}, cb::AbstractVector{Complex{T}}
     end
     Complex(re, im)
 end
+function cdot_swizzle(ca::AbstractVector{Complex{T}}, cb::AbstractVector{Complex{T}}) where {T}
+  a = reinterpret(T, ca)
+  b = reinterpret(T, cb)
+  reim = Vec(zero(T),zero(T))
+  @turbo for i ∈ eachindex(a)
+    reim = vfmsubadd(vmovsldup(a[i]), b[i], vfmsubadd(vmovshdup(a[i]), vpermilps177(b[i]), reim))
+  end
+  Complex(reim(1), reim(2))
+end
 function cdot_affine(ca::AbstractVector{Complex{T}}, cb::AbstractVector{Complex{T}}) where {T}
     a = reinterpret(T, ca);
     b = reinterpret(T, cb);
@@ -135,7 +144,7 @@ function cmatmul_array_v2!(Cc::AbstractMatrix{Complex{T}}, Ac::AbstractMatrix{Co
   C = reinterpret(Float64, Cc);
   A = reinterpret(Float64, Ac);
   B = reinterpret(reshape, Float64, Bc);
-  @turbo for n ∈ indices((C,B),(2,3)), m ∈ indices((C,A),1)
+  @turbo vectorize=2 for n ∈ indices((C,B),(2,3)), m ∈ indices((C,A),1)
     Cmn = zero(T)
     for k ∈ indices((A,B),(2,2))
       Amk = A[m,k]

--- a/test/special.jl
+++ b/test/special.jl
@@ -334,7 +334,17 @@
         end
         a
     end
-    
+
+  function sin_sum_3loop!(u, x, y, z)
+    @turbo for k in 1:length(z)
+      for j in 1:length(y)
+        for i in 1:length(x)
+          u[i, j, k] = sin(x[i]) + sin(y[j]) + sin(z[k])
+        end 
+      end 
+    end 
+  end
+
     for T ∈ (Float32, Float64)
         @show T, @__LINE__
         a = randn(T, 127);
@@ -410,6 +420,14 @@
         @test csetanh!(Y1, X, Z) ≈ csetanhavx!(Y2, X, Z)
 
         x = rand(T, 97);
-        @test transposedvectoraccessavx(x) ≈ transposedvectoraccess(x)
+      @test transposedvectoraccessavx(x) ≈ transposedvectoraccess(x)
+
+      itot = 47;
+      dx = 1. / itot;
+      x = dx*collect(0:itot-1); y = dx*collect(0:itot-1); z = dx*collect(0:itot-1);
+      u = zeros(itot+8, itot+8, itot+8);
+      uv = @view u[5:5+itot-1, 5:5+itot-1, 5:5+itot-1];
+      sin_sum_3loop!(uv, x, y, z);
+      @test uv ≈ (identity(sin.(x)) .+ identity((sin.(y))')) .+ identity(reshape(sin.(z), (1, 1, length(z))))
     end
 end

